java/ql/src/Security/CWE/CWE-020/UntrustedDataToExternalAPI.qhelp

```diff
@@ -0,0 +1,59 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+<overview>
+<p>Using unsanitized untrusted data in an external API can cause a variety of security issues. This query reports 
+all uses of external APIs with untrusted data for review. This query has a deliberately low true positive rate, 
+and is designed to help security reviews for the application, as well as helping identify external APIs that 
+should be modeled as either taint steps, or sinks for specific problems.</p>
+
+<p>An external API is defined as a method call to a method that is not defined in the source code, not overridden 
+in the source code, and is not modeled as a taint step in the default taint library. External APIs may be from the
+Java standard library, third party dependencies or from internal dependencies. The query will report uses of 
+untrusted data in either the qualifier or as one of the arguments of external APIs.</p>
+
+</overview>
+<recommendation>
+
+<p>For each result:</p>
+
+<ul>
+  <li>If the result highlights a known sink, confirm that the result is reported by the relevant query, or 
+  that the result is a false positive due to sanitization.</li>
```

Comment:
Small suggestion for clarification.

```suggestion
  that the result is a false positive because this data is sanitized.</li>
```

Question: If it is a false positive - does that suggest that this query needs attention to add the missing sanitization? Is there something the user should do?

Reply:
> Question: If it is a false positive - does that suggest that this query needs attention to add the missing sanitization? Is there something the user should do?

Not necessarily. The point I was trying to make was that if the relevant query _does not_ report this as a vulnerability, the user should confirm that some valid sanitization is preventing it from appearing. For example, if we had an XSS vulnerability reported by this query, but not the XSS query, then we should check the path for something like `sanitizeHTML`.

This query is really geared around finding false negatives, so this step is about confirming that the result is not a false negative for the relevant query. It may still be a false positive, but that's a separate issue not covered by this query or debugging process.